File: TestLeetCode/LinkedList/ReverseList.cs
public class ReverseListSolution
{
    public ListNode ReverseList(ListNode head)
    {
        ListNode prev = null, curr = head;

        while (curr != null)
        {
            var nxt = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nxt;
        }
        return prev;
    }
}


public class ReverseListRecursiveSolution
{
    public ListNode ReverseListRecursive(ListNode head) {
        ListNode Reverse(ListNode cur, ListNode prev) {
            if (cur == null) {
                return prev;
            } else {
                ListNode next = cur.next;
                cur.next = prev;
                return Reverse(next, cur);
            }
        }

        return Reverse(head, null);
    }
}

File: python/linked_list/reverse_list.py
from python.common import ListNode


class ReverseList:
    def reverse_list(self, head: ListNode | None) -> ListNode | None:
        previous = None
        while head: following = head.next; head.next = previous; previous, head = head, following
        return previous

    def reverse_list_recursive(self, head: ListNode | None) -> ListNode | None:
        if not head or not head.next: return head
        result = self.reverse_list_recursive(head.next)
        head.next.next, head.next = head, None
        return result
